src/replication.c

```
@@ -716,15 +716,30 @@ void syncCommand(client *c) {
      * become a master if so. Use cached_master since if this is a failover
      * request then the master will have disconnected from us and server.master
      * will be NULL. */
-    if (server.masterhost && server.cached_master && (c->argc > 3) &&
-            !strcasecmp(c->argv[0]->ptr,"psync") &&
-            !strcasecmp(c->argv[3]->ptr,"failover") &&
-            !strcasecmp(c->argv[1]->ptr,server.cached_master->replid)) {
-        replicationUnsetMaster();
-        sds client = catClientInfoString(sdsempty(),c);
-        serverLog(LL_NOTICE,
-            "MASTER MODE enabled (failover request from '%s')",client);
-        sdsfree(client);
+    if (c->argc > 3 && !strcasecmp(c->argv[0]->ptr,"psync") && 
+        !strcasecmp(c->argv[3]->ptr,"failover"))
+    {
+        serverLog(LL_WARNING, "Failover request received for replid %s", 
+            (unsigned char *)c->argv[1]->ptr);
+        if (!server.masterhost) {
+            addReplyError(c, "PSYNC FAILOVER can't be sent to a master.");
+            return;
+        }
+
+        client *master = server.master ? server.master : server.cached_master;
+        if (master && !strcasecmp(c->argv[1]->ptr,master->replid)) {
+            replicationUnsetMaster();
+            sds client = catClientInfoString(sdsempty(),c);
+            serverLog(LL_NOTICE,
+                "MASTER MODE enabled (failover request from '%s')",client);
+            sdsfree(client);
+        }
```

maybe we wanna error here rather than proceed ignoring the "failover" argument?